tests/moto.py

```diff
@@ -0,0 +1,101 @@
+from __future__ import annotations
+
+import time
+from collections.abc import Generator
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Any
+
+import boto3
+import pytest
+import snowflake.connector
+from moto.server import ThreadedMotoServer
+
+if TYPE_CHECKING:
+    from _pytest.scope import _ScopeName
+
+
+@dataclass
+class Session:
+    session: boto3.Session
+    endpoint_url: str
+    access_key_id: str
+    secret_access_key: str
+    session_token: str
+    region: str
+
+    def client(self, **kwargs: Any):
+        return self.session.client("s3", endpoint_url=self.endpoint_url, **kwargs)
+
+    def resource(self, **kwargs: Any):
+        return self.session.resource("s3", endpoint_url=self.endpoint_url, **kwargs)
+
+
+def create_moto_fixture(region: str = "us-east-1"):
+    @pytest.fixture()
+    def _fixture(moto_server: str, conn: snowflake.connector.SnowflakeConnection) -> Generator[Session, None, None]:
+        # Attempt at a cross-process way of generating unique 12-character integers.
+        account_id = str(time.time_ns())[:12]
+
+        sts = boto3.client(
+            "sts",
+            endpoint_url=moto_server,
+            aws_access_key_id="test",
+            aws_secret_access_key="test",
+            region_name=region,
+        )
+        response = sts.assume_role(
+            RoleArn=f"arn:aws:iam::{account_id}:role/my-role",
+            RoleSessionName="test-session-name",
+            ExternalId="test-external-id",
+        )
+
+        credentials = response["Credentials"]
+
+        access_key_id = credentials["AccessKeyId"]
+        secret_access_key = credentials["SecretAccessKey"]
+        session_token = credentials["SessionToken"]
+        with conn.cursor() as cursor:
+            cursor.execute(
+                f"""
+                CREATE SECRET s3_secret (
+                    TYPE s3,
+                    ENDPOINT '{moto_server.removeprefix("http://")}',
+                    KEY_ID '{access_key_id}',
+                    SECRET '{secret_access_key}',
+                    SESSION_TOKEN '{session_token}',
```

Comment:
Does it work without these? Might simplify things if it does.

Reply:
i dont think so. the point of this being how it is, is that otherwise by default all requests go to the same stateful moto account. whereas with this, every test gets a unique empty account that's guaranteed distinct from other tests.

there's presumably some way of resetting the state between tests within an account, but i dont like (personally) precluding the eventuality that you'd run your tests in parallel and have issues with tests clobbering one another.

Reply:
Ok, let's keep it. I have some ideas for how to hide the `Session` dataclass from the tests, and just expose an `s3_client` to them to simplify usage. And also make the moto fixtures more generic, and usable outside of this project, by moving the duckdb secret creation out. See fa21f30 which I pushed to your branch.